c/meterpreter/source/extensions/stdapi/server/net/config/route.c

```diff
@@ -229,26 +233,49 @@ DWORD add_remove_route(Packet *packet, BOOLEAN add)
 	route.dwForwardDest    = inet_addr(subnet);
 	route.dwForwardMask    = inet_addr(netmask);
 	route.dwForwardNextHop = inet_addr(gateway);
-	route.dwForwardType    = 4; // Assume next hop.
-	route.dwForwardProto   = 3;
+	route.dwForwardType    = MIB_IPROUTE_TYPE_INDIRECT; // Assume next hop.
+	route.dwForwardProto   = MIB_IPPROTO_NETMGMT;
 	route.dwForwardAge     = -1;
+	route.dwForwardMetric1 = 0;
 
-	if ((LocalGetBestInterface = (DWORD (WINAPI *)(IPAddr, LPDWORD))GetProcAddress(
-			GetModuleHandle("iphlpapi"),
-			"GetBestInterface")))
-	{
-		DWORD result = LocalGetBestInterface(route.dwForwardDest,
-				&route.dwForwardIfIndex);
+	pGetBestInterface = (GETBESTINTERFACE)GetProcAddress(GetModuleHandle(TEXT("iphlpapi")), "GetBestInterface");
+	if (!pGetBestInterface) {
+		dprintf("[NET] add_remove_route: GetBestInterface is not available.");
+		return ERROR_NOT_SUPPORTED;
+	}
 
-		if (result != ERROR_SUCCESS)
-			return result;
+	dwResult = pGetBestInterface(route.dwForwardNextHop, &route.dwForwardIfIndex);
+	if (dwResult != ERROR_SUCCESS) {
+		dprintf("[NET] add_remove_route: GetBestInterface failed. error=%d (0x%x)", dwResult, (ULONG_PTR)dwResult);
+		return dwResult;
+	}
+	dprintf("[NET] add_remove_route: GetBestInterface returned ifIndex=%d.", route.dwForwardIfIndex);
+
+	pGetIpInterfaceEntry = (GETIPINTERFACEENTRY)GetProcAddress(GetModuleHandle(TEXT("iphlpapi")), "GetIpInterfaceEntry");
+	// If GetIpInterfaceEntry is available, use it to set the default metric because newer systems require that
+	if (pGetIpInterfaceEntry) {
+		MIB_IPINTERFACE_ROW iface = { .Family = AF_INET, .InterfaceIndex = route.dwForwardIfIndex };

```

Comment:
Curious but why the dot before these attributes? Those don't seem to be part of the name. Also should we not be zeroing out the memory of this structure before assigning values? Otherwise we are relying on the compiler doing this or stuff to not be used.

Reply:
> Curious but why the dot before these attributes? Those don't seem to be part of the name.

They are names in the [MIB_IPINTERFACE_ROW](https://learn.microsoft.com/en-us/windows/win32/api/netioapi/ns-netioapi-mib_ipinterface_row) definition. If you remove the leading dot the compiler breaks throwing the following error:

![image](https://user-images.githubusercontent.com/2058303/222808227-6d1de55c-345c-44a6-9158-cd07522b2bb2.png)

> Also should we not be zeroing out the memory of this structure before assigning values? Otherwise we are relying on the compiler doing this or stuff to not be used.

We covered this [in the last PR.](https://github.com/rapid7/metasploit-payloads/pull/599#discussion_r1070091489) Per the C-99 spec, a partial definition such as this will cause the other fields to be initialized to zero. It's dependent on the compiler I guess, but in the sense that the compiler needs to conform to the spec.

Reply:
My apologies on the second point; technically speaking its a little bit of a shakey way of doing things but since we are only compiling with one compiler (VS) on one specific spec, this shouldn't be an issue if the spec defines that it should be initialized. Thanks for clarifying this.

As for the other point, appreciate you sharing that, makes sense.

Based on this I'm going to mark this comment as resolved.